In [9]:
import cv2
import numpy as np
import sys

마스크 연산을 이용하여 이미지 합성하는 방법


In [10]:
src = cv2.imread('airplane.bmp', cv2.IMREAD_COLOR)
mask = cv2.imread('mask_plane.bmp', cv2.IMREAD_GRAYSCALE)

dst = cv2.imread('field.bmp', cv2.IMREAD_COLOR)

필요한 image data들 불러옴

In [11]:
cv2.copyTo(src, mask, dst)
#dst = cv2.copyTo(src, mask)


array([[[203, 199, 207],
        [199, 200, 208],
        [202, 201, 211],
        ...,
        [198, 190, 194],
        [195, 188, 190],
        [193, 184, 189]],

       [[203, 199, 207],
        [204, 198, 207],
        [200, 199, 209],
        ...,
        [173, 157, 148],
        [170, 156, 143],
        [184, 173, 166]],

       [[203, 196, 208],
        [204, 196, 208],
        [203, 198, 208],
        ...,
        [152, 129, 109],
        [151, 121, 108],
        [148, 126,  97]],

       ...,

       [[ 51,  29,  15],
        [ 49,  32,  29],
        [ 49,  49,  61],
        ...,
        [ 45,  62,  38],
        [ 46,  75,  56],
        [ 40, 110, 100]],

       [[ 50,  26,  11],
        [ 50,  45,  54],
        [ 47,  44,  50],
        ...,
        [ 49,  56,  35],
        [ 47,  61,  42],
        [ 44,  77,  64]],

       [[ 50,  42,  52],
        [ 46,  71, 108],
        [ 47,  56,  58],
        ...,
        [ 50,  42,  23],
        [ 48,  61,  41],
        [ 47,  62,  46]]

src, dst 는 color type이 같아야 하고 mask는 무조건 grayscale

copyto에 src, mask, dst가 모두 인가 된다면 'src' 중 'masking' 된 부분이 겹쳐지면서 올라간다.

아니라면 'src' 중 'masking'으로 분리된 부분만 출력된다.

In [12]:
dst[mask > 0] = src[mask > 0]

또 다른 방법에는 mask의 영역값(인덱스들)과 동일한 부분의 src영역의 값들을 dst의 동일한 영역에 넣어줄 수있다.

이 방법은 copy가 아닌 동일한 메모리에 있는 정보에 인덱싱 하는것이라

dst 파일 자체의 값이 변하게 된다.

In [13]:
cv2.imshow('src', src)
cv2.imshow('mask', mask)
cv2.imshow('dst', dst)

cv2.waitKey()
cv2.destroyAllWindows()

In [14]:
src = cv2.imread('opencv-logo-white.png', cv2.IMREAD_UNCHANGED)
print(src.shape)

(222, 180, 4)


새롭게 src지정, 이 파일은 4ch로써 마지막에 alpha channel을 가지게 된다.
(222, 180, 4)

In [15]:
mask = src[:, :, -1]
src = src[:, :, 0:3]
dst = cv2.imread('field.bmp', cv2.IMREAD_COLOR)

h, w = src.shape[:2] 
dst_h, dst_w = dst.shape[:2]
crop = dst[(dst_h-h):(dst_h), (dst_w-w):(dst_w)]

src의 마지막 alpha ch을 mask기준값으로 사용하고

마지막 4채널을 제외한 color이미지를 src이미지로 사용한다.

copyto 를 통해 이미지를 합성하기 위해서 입력 이미지들의 크기가 동일해야한다. 

crop필요하다. 일부분의 data를 crop이란 변수로 공유한다.



In [19]:
cv2.copyTo(src, mask, crop)
cv2.imshow('crop', dst)

cv2.waitKey()
cv2.destroyAllWindows()

crop을 통해 dst의 일부를 참조하였고 그 data로 copyto를 진행하였기 때문에

결국에 dst까지도 변경되어서 최종출력을 dst로 해도 가능하다